// Copyright 2019 The GraphicsFuzz Project Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package gfauto;

import "gfauto/common.proto";


// Artifact metadata.
message ArtifactMetadata {

  // The artifact's type-specific data.
  Data data = 1;

  // A list of artifact paths from which this artifact is derived. Typically, this is similar to
  // following the input artifacts in the recipe transitively, but some artifacts may not have a
  // recipe because they were created manually so it is still useful to have this field.
  repeated string derived_from = 2;

  string comment = 3;

  message Data {
    oneof data {
      ArtifactMetadataGlslShaderJob glsl_shader_job = 1;
      ArtifactMetadataGlslReduction glsl_shader_job_reduction = 2;
      ArtifactMetadataSpirvShaderJob spirv_shader_job = 3;
      ArtifactMetadataSpirvAsmShaderJob spirv_asm_shader_job = 4;
      ArtifactMetadataAmberScript amber_script = 5;
      ArtifactMetadataTextFile text_file = 6;
      ArtifactMetadataExtractedArchiveSet extracted_archive_set = 7;
    }
  }
}


message ArtifactMetadataGlslShaderJob {
  string shader_job_file = 1;
  string reference_glsl_shader_job_artifact = 2;
  bool red_pixel_at_top_left = 3;
  // TODO: Consider having nested message for ImageOutput (reference info, red pixel info, etc.) and
  //  BufferOutput.
}

message ArtifactMetadataGlslReduction {
  string final_shader = 1;
}

// Common metadata about a spirv shader job, be it binary spirv or spirv assembly.
message MetadataSpirvJob {
  string shader_job_file = 1;
  string glsl_shader_job_source_artifact = 2;

  // Indicates the spirv-opt args that were used when processing this shader. The recipe gives the
  // the full story.
  repeated string spirv_opt_args = 3;

  // Is the top-left pixel guaranteed to be a red pixel?
  bool red_pixel_at_top_left = 4;

  // Flag to indicate that the SPIR-V was generated but then manually modified.
  bool manually_modified = 5;
}

message ArtifactMetadataSpirvShaderJob {
  MetadataSpirvJob spirv_job = 1;
}

message ArtifactMetadataSpirvAsmShaderJob {
  MetadataSpirvJob spirv_job = 1;
}

message ArtifactMetadataAmberScript {
  string amber_script_file = 1;
  bool self_contained_test = 2;
  string glsl_shader_job_source_artifact = 3;
}

message ArtifactMetadataTextFile {
  string text_file = 1;
}

message ArtifactMetadataExtractedArchiveSet {
  ArchiveSet archive_set = 1;
}
